base/services/w32time/ntpclient.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * PROJECT:     ReactOS Timedate Control Panel
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     Queries the NTP server
+ * COPYRIGHT:   Copyright 2006 Ged Murphy <gedmurphy@gmail.com>
+ */
+
+#include "timedate.h"
+#include <winsock2.h>
+
+#define TIMEOUT 4000 /* 4 second timeout */
+
+typedef struct _INFO
+{
+    SOCKET Sock;
+    SOCKADDR_IN myAddr;
+    SOCKADDR_IN ntpAddr;
+    NTPPACKET SendPacket;
+    NTPPACKET RecvPacket;
+} INFO, *PINFO;
+
+
+static BOOL
+InitConnection(PINFO pInfo, LPSTR lpAddress)
+{
+    WSADATA wsaData;
+    HOSTENT *he;
+    INT Ret;
+
+    Ret = WSAStartup(MAKEWORD(2, 2), &wsaData);
+    if (Ret != 0)
+        return FALSE;
+
+    pInfo->Sock = socket(AF_INET,
+                         SOCK_DGRAM,
+                         0);
+    if (pInfo->Sock == INVALID_SOCKET)
+        return FALSE;
+
+    /* Setup server info */
+    he = gethostbyname(lpAddress);
+    if (he != NULL)
+    {
+        /* Setup server socket info */
+        ZeroMemory(&pInfo->ntpAddr, sizeof(SOCKADDR_IN));
+        pInfo->ntpAddr.sin_family = AF_INET; // he->h_addrtype;
+        pInfo->ntpAddr.sin_port = htons(NTPPORT);
+        pInfo->ntpAddr.sin_addr = *((struct in_addr *)he->h_addr);
+    }
+    else
+        return FALSE;
+
+    return TRUE;
+}
+
+
+static VOID
+DestroyConnection(VOID)
+{
+    WSACleanup();
+}
+
+
+static BOOL
+GetTransmitTime(PTIMEPACKET ptp)
+{
+    return TRUE;
+}
+
+
+/* Send some data to wake the server up */
+static BOOL
+SendData(PINFO pInfo)
+{
+    TIMEPACKET tp = { 0, 0 };
+    INT Ret;
+
+    ZeroMemory(&pInfo->SendPacket, sizeof(pInfo->SendPacket));
+    pInfo->SendPacket.LiVnMode = 27;
+    if (!GetTransmitTime(&tp))
+        return FALSE;
+    pInfo->SendPacket.TransmitTimestamp = tp;
+
+    Ret = sendto(pInfo->Sock,
+                 (char *)&pInfo->SendPacket,
+                 sizeof(pInfo->SendPacket),
+                 0,
+                 (SOCKADDR *)&pInfo->ntpAddr,
+                 sizeof(SOCKADDR_IN));
```

Comment:
Shouldn't it be sizeof(pInfo->ntpAddr) ? (of course in case this ntpAddr variable is of SOCKADDR_IN type).

Reply:
I do not know. This was copied from dll/cpl/timedate/ntpclient.c by Ged Murphy. Thanks.

Reply:
I see this duplicates the entire file. It would be better to move it to a library.